### PR TITLE
Enable DI for storage session

### PR DESCRIPTION
**Pull request by SzymonPobiega**

### CODE REVIEW

File: src/NServiceBus.NHibernate.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt [hunk 1]
```diff
@@ -25,6 +25,23 @@ namespace NServiceBus.Features
         protected override void Setup(NServiceBus.Features.FeatureConfigurationContext context) { }
     }
 }
+namespace NServiceBus
+{
+    public interface INHibernateStorageSession
+    {
+        NHibernate.ISession Session { get; }
+        void OnSaveChanges(System.Func<NServiceBus.Persistence.SynchronizedStorageSession, System.Threading.Tasks.Task> callback);
```
Comment (tmasternak) [inline src/NServiceBus.NHibernate.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt:33]: Is the `OnSaveChanges` added for completness?

File: src/NServiceBus.NHibernate.AcceptanceTests/When_using_synchronized_session_via_container.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                {
+                    c.RegisterComponents(cc =>
+                    {
+                        cc.ConfigureComponent<MyRepository>(DependencyLifecycle.InstancePerUnitOfWork);
+                        cc.ConfigureComponent(b => b.Build<INHibernateStorageSession>().Session, DependencyLifecycle.InstancePerUnitOfWork);
+                    });
+                });
+            }
+
+            public class MyMessageHandler : IHandleMessages<MyMessage>
```
Comment (tmasternak) [inline src/NServiceBus.NHibernate.AcceptanceTests/When_using_synchronized_session_via_container.cs:45]: Should we add another handler and test that session is the same for the whole `UnitOfWork`?

File: src/NServiceBus.NHibernate.AcceptanceTests/When_using_synchronized_session_via_container.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+namespace NServiceBus.AcceptanceTests.Sagas
+{
+    using global::NHibernate;
+
+    [TestFixture]
+    public class When_using_synchronized_session_via_container : NServiceBusAcceptanceTest
+    {
+        [Test]
+        public async Task Should_inject_synchronized_session_into_handler()
```
Comment (tmasternak) [inline src/NServiceBus.NHibernate.AcceptanceTests/When_using_synchronized_session_via_container.cs:14]: Can we test `OnSaveChanges`

File: src/NServiceBus.NHibernate.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt (hunk 1 above)
Comment (SzymonPobiega) [reply src/NServiceBus.NHibernate.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt:33]: It was always there but I changed the name of the interface from `INHibernateSynchronizedStorageSession` to `INHibernateStorageSession` when making it public. Before this PR, the API was available via extension methods on `SynchronizedStorageSessionExtensions`.

File: src/NServiceBus.NHibernate.AcceptanceTests/When_using_synchronized_session_via_container.cs (hunk 3 above)
Comment (SzymonPobiega) [reply src/NServiceBus.NHibernate.AcceptanceTests/When_using_synchronized_session_via_container.cs:14]: It is tested in the unit tests e.g. `NHibernateAmbientTransactionSynchronizedStorageSessionTests` and two others in the same folder.

File: src/NServiceBus.NHibernate.AcceptanceTests/When_using_synchronized_session_via_container.cs (hunk 2 above)
Comment (SzymonPobiega) [reply src/NServiceBus.NHibernate.AcceptanceTests/When_using_synchronized_session_via_container.cs:45]: Would it be enough to call `GetHashCode` and ensure both sessions are actually the same instance?

Comment (tmasternak) [reply src/NServiceBus.NHibernate.AcceptanceTests/When_using_synchronized_session_via_container.cs:45]: I was thinking about this case a there were some workarounds needed (asynclocal) to make this work. And yes I think that `GetHashCode` is enough.

**Pull request merged**